### PR TITLE
[Merged by Bors] - make ComponentTicks::set_changed public

### LINKED ISSUE #1710: Ability to dynamically mark components as changed
Before the reliable change pr one could insert `ComponentFlags::MUTATED` to mark a component as changed in a context where the type isn't known at compile time.

The equivalent function `ComponentTicks::set_changed` is now `pub(crate)`:
https://github.com/bevyengine/bevy/blob/45b2db70705da24a89426e6b6e77d603a3983025/crates/bevy_ecs/src/component/mod.rs#L344
The simple fix would be to make that function `pub`.

Alternatively we could expose a version of `EntityMut::get_mut` which passes in the component id:
```rust
#[inline]
pub fn get_mut_dynamic(&mut self, component_id: ComponentId) -> Option<MutDynamic<'w>> {
   // ...
}
```

### PR DESCRIPTION
**Pull request by jakobhellermann**
fixes #1710

### CODE REVIEW

**Review by Davier — APPROVED** [review]

**Comment by mockersf** [comment]
for public consumption, could we have another method that takes a `bool` for `set_changed`?

**Comment by jakobhellermann** [comment]
It would also need to take the world for getting the current change tick, right? I haven't been following the reliable change detection work much.

**Comment by mockersf** [comment]
I didn't check either, but it looks like `0` or `u32::MAX` would be interesting values to set for "changed/didn't change"

**Comment by alice-i-cecile** [comment]
My preferred API to expose for this would be a `set_changed(bool)` method on commands (and World) that operates on a single entity, following #1703 .

As a bit of context, @cart mentioned in #1471 that the requested functionality here might be needed, but to wait for a separate pr / use case. This seems like just that :)

**Comment by alice-i-cecile** [comment]
Adding to the 0.5 milestone: I'd like to avoid a regression in useful functionality, especially for such a critical ecosystem plugin.

**Comment by alice-i-cecile** [comment]
My concern with exposing the API directly as done in the first commit is that it's very, very coupled to the implementation details. This makes it very liable to break, and forces a huge burden of knowledge on the end user to use correctly.

**Comment by Davier** [comment]
I'm against `set_changed(bool)` and/or `set_added(bool)`. It makes no sense to me to call them with `false`. What would be the use cases?

>My concern with exposing the API directly as done in the first commit is that it's very, very coupled to the implementation details. This makes it very liable to break, and forces a huge burden of knowledge on the end user to use correctly.

`ComponentTicks::set_changed` is not something that should generally be used, but making it public for corner cases like described in #1710 seems ok to me. It could use a doc comment though. The alternative is to be forced to use `ReflectMut` (which is the intended way to do it, so I'd be ok with that too).

What cart suggested in #1471 is to have `commands.set_changed::<T>(entity)` to be able to set components as mutated without having mutable access to them, which is a different issue and doesn't solve "setting mutated for components of unknown types". I think that should be its own PR, and for reference my preferred solution is to add `set_changed` in `ChangeTrackers<T>` rather than a command.

**Comment by mockersf** [comment]
rather than having a `set_changed(bool)`, what I would like to avoid is having a `set_changed(u32)` public as that may be very error prone if we let the user give whatever u32 they want...

**Comment by Davier** [comment]
> rather than having a `set_changed(bool)`, what I would like to avoid is having a `set_changed(u32)` public as that may be very error prone if we let the user give whatever `u32` they want...

The issue as I understand it is that `ReflectMut` can't be used in `bevy_inspector_egui` because they have a custom reflection mechanism, so they want access to this low-level API instead. I'm not too worried about exposing it since you need to use some low-level `unsafe` API to even get a `ComponentTicks` struct, it's unlikely to be used by accident.

**Comment by alice-i-cecile** [comment]
> I'm against set_changed(bool) and/or set_added(bool). It makes no sense to me to call them with false. What would be the use cases?

Hmm. I can imagine specialized use cases (e.g. specialized loading workflows) where we set these as false, but they would be both obscure and dangerous. Probably better to avoid encouraging it.

> I'm not too worried about exposing it since you need to use some low-level unsafe API to even get a ComponentTicks struct, it's unlikely to be used by accident.

I can accept this,  provided the doc strings are clear that it's probably not what you want to be using and we eventually create commands to set entities as added / changed in a user-friendly way.

**Comment by jakobhellermann** [comment]
> I can accept this, provided the doc strings are clear that it's probably not what you want to be using and we eventually create commands to set entities as added / changed in a user-friendly way.

I added a doc comment explaining how this is usually done automatically.

**Review by alice-i-cecile — APPROVED** [review]

File: crates/bevy_ecs/src/component/mod.rs [hunk 1]
```diff
@@ -340,8 +340,19 @@ impl ComponentTicks {
         check_tick(&mut self.changed, change_tick);
     }
 
+    /// Manually sets the change tick.
+    /// Usually, this is done automatically via the [`Deref`](std::ops::Deref) implementation on [`Mut`](crate::world::Mut) or [`ResMut`](crate::system::ResMut) etc.
```
Comment (Davier) [inline crates/bevy_ecs/src/component/mod.rs:344]: It's only done in `DerefMut`, not `Deref`.

File: crates/bevy_ecs/src/component/mod.rs [hunk 2]
```diff
@@ -340,8 +340,19 @@ impl ComponentTicks {
...
+    ///
+    /// # Example
+    /// ```rust,no_run
+    /// # use bevy_ecs::{world::World, component::ComponentTicks};
+    /// let world: World = todo!();
+    /// let component_ticks: ComponentTicks = todo!();
```
Comment (Davier) [inline crates/bevy_ecs/src/component/mod.rs:350]: Do you intend to fill this in this PR?

File: crates/bevy_ecs/src/component/mod.rs (hunk 1 above)
Comment (jakobhellermann) [reply crates/bevy_ecs/src/component/mod.rs]: Fixed.

File: crates/bevy_ecs/src/component/mod.rs (hunk 2 above)
Comment (jakobhellermann) [reply crates/bevy_ecs/src/component/mod.rs:350]: I just wanted the variables in scope, but I agree that `todo!()` doesn't really fit. Do you think changing it to `unimplemented!()` would be fine or would you like a larger example with proper values?

Its hard to come up with a simple example as this is the code I used (copied from bevy):
```rust
#[inline]
unsafe fn get_component_and_ticks(
    world: &World,
    component_id: ComponentId,
    entity: Entity,
    location: EntityLocation,
) -> Option<(*mut u8, *mut ComponentTicks)> {
    let archetype = &world.archetypes()[location.archetype_id];
    let component_info = world.components().get_info_unchecked(component_id);
    match component_info.storage_type() {
        StorageType::Table => {
            let table = &world.storages().tables[archetype.table_id()];
            let components = table.get_column(component_id)?;
            let table_row = archetype.entity_table_row(location.index);
            // SAFE: archetypes only store valid table_rows and the stored component type is T
            Some((
                components.get_unchecked(table_row),
                components.get_ticks_unchecked(table_row),
            ))
        }
        StorageType::SparseSet => world
            .storages()
            .sparse_sets
            .get(component_id)
            .and_then(|sparse_set| sparse_set.get_with_ticks(entity)),
    }
}
```

Comment (cart) [reply crates/bevy_ecs/src/component/mod.rs:350]: Yeah lets just change these to unimplemented().

**Comment by cart** [comment]
I'm cool with just making this public. ChangedTicks is "just" a holder of the last changed ticks. Adding a `set_changed(bool)` or `set_changed()` method feels like a higher level construct better suited to `Mut` pointers or something like `ChangeTrackersMut`..

File: crates/bevy_ecs/src/component/mod.rs (hunk 2 above)
Comment (jakobhellermann) [reply crates/bevy_ecs/src/component/mod.rs]: Done

**Pull request closed without merging**